xswap/xswap_backend.cpp

```diff
@@ -0,0 +1,194 @@
+    EdgeHashTable edges_set = EdgeHashTable(max_source, max_target);
+    for (int i = 0; i < num_edges; i++) {
+        edges_set.add(edges[i]);
+    }
+
+    // Initialize unbiased random number generator
+    std::mt19937 rng(seed);
+    std::uniform_int_distribution<int> uni(0, num_edges - 1);
+
+    // Initialize stats counters for failure reasons
+    int same_edge = 0;
+    int self_loop = 0;
+    int duplicate = 0;
+    int undir_duplicate = 0;
+    int excluded = 0;
+
+    // Do XSwap
+    for (int i = 0; i < num_swaps; i++) {
+        // Draw edges randomly
+        int edge_index_a = uni(rng);
+        int edge_index_b = uni(rng);
+
+        if (edge_index_a == edge_index_b) {
+            same_edge += 1;
+            continue;
+        }
+
+        // Old edges
+        int* edge_a = edges[edge_index_a];
+        int* edge_b = edges[edge_index_b];
+
+        // Form potential new edges
+        int new_edge_a[2] = { edge_a[0], edge_b[1] };
+        int new_edge_b[2] = { edge_b[0], edge_a[1] };
+        int* new_edges[2] = { new_edge_a, new_edge_b };
+        // Check validity of both new edges
+        bool valid = true;
+        for (int i = 0; i < 2; i++) {
+            int* new_edge = new_edges[i];
+            // New edge would be a self-loop
+            if (new_edge[0] == new_edge[1]) {
+                self_loop += 1;
+                valid = false;
+                break;
+            }
+            // New edge already exists
+            if (edges_set.contains(new_edge)) {
+                duplicate += 1;
+                valid = false;
+                break;
+            }
+            // Undirected and reverse of new edge already exists
+            int reversed[2] = { new_edge[1], new_edge[0] };
+            if (!directed && edges_set.contains(reversed)) {
+                undir_duplicate += 1;
+                valid = false;
+                break;
+            }
+            // New edges are in excluded edges. Because few excluded edges are expected,
+            // this iterates in O(n) rather than allocating a large hash table for all
+            // possible edges. Considerable memory savings when excluding few edges.
+            // An alternative is to pass excluded edges as a Python set and to use the
+            // built-in PySet_Contains function to check membership. This has the downside
+            // that the edge must be instantiated as a new Python tuple object and the
+            // values represented as new Python long integer objects.
+            for (int i = 0; i < num_excluded_edges; i++) {
+                if (excluded_edges[i][0] == new_edge[0] && excluded_edges[i][1] == new_edge[1]) {
+                    excluded += 1;
+                    valid = false;
+                    break;
+                }
+            }
+        }
+        if (valid) {
+            edges_set.remove(edge_a);
+            edges_set.remove(edge_b);
+
+            int temp_target = edge_a[1];
```

Comment:
If you want, you can use `std::swap(edge_a[1], edge_b[1])`, but you have to `#include <algorithm>`.